R/epidatacall.R

```diff
@@ -68,187 +58,219 @@ create_epidata_call <- function(endpoint, params, meta = NULL,
-      if (info$name %in% columns && info$type == "epiweek") {
-        tbl[[info$name]] <- parse_api_week(tbl[[info$name]])
-      }
-    }
-  }
-  tbl
+  df <- fetch_classic(epidata_call, fields, disable_data_frame_parsing = FALSE)$epidata
+  return(parse_data_frame(epidata_call, df, disable_date_parsing) %>% as_tibble())
 }
 
-#' Fetches the data and returns the classic format
+
+
+#' Fetches the data and returns a JSON object in the classic format
+#'
+#' The classic format is a 3-element list with the following elements:
+#'   - `result`: a status code (1 for success, -2 for no-results, all else for failure)
+#'   - `message`: a string describing the status of the request
+#'   - `epidata`: a list of named lists representing rows of a table, or, for a
+#'                few endpoints, a more complex nested list structure
+#'
 #' @rdname epidata_call
 #'
 #' @param epidata_call an instance of `epidata_call`
 #' @param fields filter fields
-#' @param disable_date_parsing disable automatic date parsing
+#' @param disable_data_frame_parsing do not automatically cast the epidata output to a data frame
+#'   (some endpoints return a list of lists, which is not a data frame)
 #' @importFrom httr stop_for_status content http_error
 #' @importFrom jsonlite fromJSON
 #' @return
-#' - For `fetch_classic`: the "classic" 3-element list format (see above)
+#' - For `fetch_classic`: a JSON-like list
 #'
 #' @export
-fetch_classic <- function(epidata_call, fields = NULL, disable_date_parsing = FALSE) {
+fetch_classic <- function(epidata_call, fields = NULL, disable_data_frame_parsing = TRUE) {
   stopifnot(inherits(epidata_call, "epidata_call"))
   stopifnot(is.null(fields) || is.character(fields))
-  stopifnot(is.logical(disable_date_parsing), length(disable_date_parsing) == 1)
+
   res <- request_impl(epidata_call, "classic", fields)
-  r <- httr::content(res, "text", encoding = "UTF-8")
-  if (httr::http_error(res)) {
-    # return message in case of error
-    return(list(result = 0, message = paste0("error: ", r), epidata = data.frame()))
+  if (res$status_code != 200) {
+    # 500, 429, 401 are possible
+    msg <- "fetch data from API"
+    if (http_type(res) == "text/html") {
+      # grab the error information out of the returned HTML document
+      msg <- paste(msg, ":", xml2::xml_text(xml2::xml_find_all(
+        xml2::read_html(content(res, 'text')),
+        "//p"
+      )))
+    }
+    httr::stop_for_status(res, task = msg)
   }
+  r <- httr::content(res, "text", encoding = "UTF-8")
 
-  m <- jsonlite::fromJSON(r)
-  if ("epidata" %in% names(m) && !(epidata_call$only_supports_classic)) {
-    m$epidata <- parse_data_frame(epidata_call, m$epidata, disable_date_parsing = disable_date_parsing)
+  r <- jsonlite::fromJSON(r, simplifyDataFrame = !disable_data_frame_parsing)
+  # success is 1, no results is -2
+  if (r$result != 1 && r$result != -2) {
+    rlang::abort(paste0("epidata error: ", r$message), "epidata_error")
+  } else {
+    r
```

Comment:
question: should we `rlang::warn` if `r$message != "success"`?
question: if yes, why not just return `r$epidata` at the end of this?  We've already propagated all errors and warnings.

question: do we even want to allow partial results?  There's no good way to get the rest of the results.  I think it'd be more helpful for that to be a hard error.

Reply:
See my comment below. I'd rather not have someone call fetch_classic and get the results of fetch_json. So I'd prefer to just not raise errors at all and maybe warn instead.

As for the final question, I don't think our server notifies when results are partial. But even if it does, it seems strange to let someone download 10 million rows and then to discard the download with a "partial" error.

Reply:
I'd then want something like `fetch_json` (but without dataframeification) or the alternative name `fetch_object`, which DO stop and warn in all cases, and would advertise those over `fetch_classic`, and use them in `fetch` rather than `fetch_classic`.  Reasoning: `fetch_classic` is just asking for people to accidentally ignore errors and warnings, as we did in this very client before.  And `fetch_object` and `fetch_tbl` are more similar to each other, so they  would mix better / surprise less in `fetch()`.

Do we still have endpoints around that have a limit of 3650 rows?  Guess I have some old annoyances that might not apply now.  But why is the server doing truncation and not a hard error at 10M rows?  It's errors waiting to happen again.  I guess it's hard to know what how many rows a query will generate ahead of time, but if I were a user I'm probably not going to try to debug exactly which 10M rows I got and which I'm missing and really rely on my guesses in getting the missing rows with another (potentially very hard to write) query.  I'd dump the 10M and rewrite to fetch the entire thing in a way that wouldn't hit the limit in any individual query.

Reply:
Re: limits. Looks like people still hitting the old `api.php` endpoint still hit the 3650 row limit, I guess for compatibility reasons 🤦. Also [the server gives a status code of 2 when it hits the limit](https://github.com/cmu-delphi/delphi-epidata/blob/52f57d6cd6d3da967c24e9d2741e727560bf3e27/src/server/_printer.py#L88-L91). People using this client, unless for some reason they choose to `with_base_url` the old php endpoint, a choice that I think we can safely ignore, will have the 10e6 limit. You're right, the partial results aren't very useable, so I am fine with throwing an error if we get a truncated message. It would suck for the user to wait a long time to only get an error back, but I don't think there's much to do about that.

Re: fetch functions. I just don't see why we need to export all these `fetch_*` functions. I thought we were aligned on this PR's goal of centralizing everything around `fetch`, but I suppose not? I would prefer to delete `fetch_json`, unexport `fetch_classic` and `fetch_tbl`, and only export `fetch`. This function will then return a tibble in 99% of use cases, will raise on HTTP errors, surface HTTP body messages as warnings, will raise on epidata errors, and in the few only_supports_classic cases, will return a raw JSON object from the API call. The three remaining only_supports_classic endpoints are either barely used or are already covered - `delphi` is [broken](https://api.covidcast.cmu.edu/epidata//delphi/?system=ec&epiweek=202006), `pvt_meta_norostat` is one of the private endpoints [untouched in years](https://delphi-org.slack.com/archives/C02U0AX8KJR/p1683837318090089), and `meta` which is already covered by `covidcast_epidata()`. The classic format is explained to the user in the docstring, so they'll know what to expect from these endpoints.

Reply:
I thought the `fetch` idea here was originally to /default/ to tbl or classic, but also /allow/ other things.  As we discuss though, I'm okay with trying to eliminate some of the options.  But I don't think result-status-epidata without stop/warn should ever be the default output, and it stop&warn&return epidata should be an option for all endpoints.  Plus it seems a little simpler and less error-prone to have `fetch` to stop&warn in all cases, and return the epidata in all cases (turned into tbl for non-only-classic endpoints only).  The problem with not stopping is that people don't check the status&message and file bug reports, and the problem with not warning (larger) is that they may NEVER catch these and just have bad input & bad output.